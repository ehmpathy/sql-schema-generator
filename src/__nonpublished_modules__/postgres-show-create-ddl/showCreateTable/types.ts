import { QueryResult } from 'pg';

export interface DatabaseConnection {
  query: (args: { sql: string; values?: (string | number)[] }) => Promise<QueryResult<any>>;
  end: () => Promise<void>;
}
