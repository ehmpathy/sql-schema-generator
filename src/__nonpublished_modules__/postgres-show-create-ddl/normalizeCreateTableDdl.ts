import sqlFormatter from 'sql-formatter';

/**
 * normalize generated create table ddl for comparisons
 *
 * TODO: name this function specific to its use case
 * e.g., "normalizeForSchemaComparison" or something similar
 */
export const normalizeCreateTableDdl = ({ ddl }: { ddl: string }) => {
  // prettify that ddl, to make standard spacing and appearance of ddl
  const prettyDdl = sqlFormatter.format(ddl);

  // swap back to better aliases
  let aliasedSql = prettyDdl;
  aliasedSql = aliasedSql.replace(/int DEFAULT nextval\([\'\w: \n]+\)/g, 'serial'); // e.g., `bigint DEFAULT nextval('test_tb_for_show_create_on_id_seq' :: regclass)` -> `bigserial`; although 'bigint default ...' is the canonical def, its just too verbose to be useful
  aliasedSql = aliasedSql.replace(/character varying/g, 'varchar'); // although `character varying` is the canonical term, `varchar` is used more regularly in practice -> more helpful
  aliasedSql = aliasedSql.replace(/ integer /g, ' int '); // postgres converts `int` to `integer`, but `bigint` is kept as `bigint`.... lets be consistent and just use the shorthand
  aliasedSql = aliasedSql.replace(/ ?:: ?\w+( ?\[\])?/g, ''); // remove any ":: __TYPE__" casting that could exist in the DDL. this level of information has not been found as useful yet

  // reprettify the ddl
  const reprettyDdl = sqlFormatter.format(aliasedSql);

  // return the normalized ddl
  return reprettyDdl;
};
