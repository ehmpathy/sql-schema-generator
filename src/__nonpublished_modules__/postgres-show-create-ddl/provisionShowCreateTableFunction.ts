import { DatabaseConnection } from './types';

/**
 * fortunately for us, postgres does not come with built in `SHOW CREATE TABLE/VIEW/FUNCTION`
 *
 * we have to provide a way to do so manually. on the bright side, postgres does provide an extensive catalog system that we can leverage
 *
 * based on https://stackoverflow.com/questions/2593803/how-to-generate-the-create-table-sql-statement-for-an-existing-table-in-postgr
 */
export const provisionShowCreateTableFunction = async ({ dbConnection }: { dbConnection: DatabaseConnection }) => {
  const sql = `
CREATE OR REPLACE FUNCTION public.show_create_table(p_schema_name character varying, p_table_name character varying)
  RETURNS SETOF text AS
$BODY$
DECLARE
    v_table_ddl   text;
    column_record record;
    table_rec record;
    constraint_rec record;
    firstrec boolean;
BEGIN
    FOR table_rec IN
        SELECT c.relname, c.oid FROM pg_catalog.pg_class c
            LEFT JOIN pg_catalog.pg_namespace n ON n.oid = c.relnamespace
                WHERE relkind = 'r'
                AND n.nspname = p_schema_name
                AND relname~ ('^('||p_table_name||')$')
          ORDER BY c.relname
    LOOP
        FOR column_record IN
            SELECT
                b.nspname as schema_name,
                b.relname as table_name,
                a.attname as column_name,
                pg_catalog.format_type(a.atttypid, a.atttypmod) as column_type,
                CASE WHEN
                    (SELECT substring(pg_catalog.pg_get_expr(d.adbin, d.adrelid) for 128)
                     FROM pg_catalog.pg_attrdef d
                     WHERE d.adrelid = a.attrelid AND d.adnum = a.attnum AND a.atthasdef) IS NOT NULL THEN
                    'DEFAULT '|| (SELECT substring(pg_catalog.pg_get_expr(d.adbin, d.adrelid) for 128)
                                  FROM pg_catalog.pg_attrdef d
                                  WHERE d.adrelid = a.attrelid AND d.adnum = a.attnum AND a.atthasdef)
                ELSE
                    ''
                END as column_default_value,
                CASE WHEN a.attnotnull = true THEN
                    'NOT NULL'
                ELSE
                    'NULL'
                END as column_not_null,
                a.attnum as attnum,
                e.max_attnum as max_attnum
            FROM
                pg_catalog.pg_attribute a
                INNER JOIN
                 (SELECT c.oid,
                    n.nspname,
                    c.relname
                  FROM pg_catalog.pg_class c
                       LEFT JOIN pg_catalog.pg_namespace n ON n.oid = c.relnamespace
                  WHERE c.oid = table_rec.oid
                  ORDER BY 2, 3) b
                ON a.attrelid = b.oid
                INNER JOIN
                 (SELECT
                      a.attrelid,
                      max(a.attnum) as max_attnum
                  FROM pg_catalog.pg_attribute a
                  WHERE a.attnum > 0
                    AND NOT a.attisdropped
                  GROUP BY a.attrelid) e
                ON a.attrelid=e.attrelid
            WHERE a.attnum > 0
              AND NOT a.attisdropped
            ORDER BY a.attnum
        LOOP
            IF column_record.attnum = 1 THEN
                v_table_ddl:='CREATE TABLE '||column_record.schema_name||'.'||column_record.table_name||' (';
            ELSE
                v_table_ddl:=v_table_ddl||',';
            END IF;

            IF column_record.attnum <= column_record.max_attnum THEN
                v_table_ddl:=v_table_ddl||chr(10)||
                         '    '||column_record.column_name||' '||column_record.column_type||' '||column_record.column_default_value||' '||column_record.column_not_null;
            END IF;
        END LOOP;
        firstrec := TRUE;
        FOR constraint_rec IN
            SELECT conname, pg_get_constraintdef(c.oid) as constrainddef
                FROM pg_constraint c
                    WHERE conrelid=(
                        SELECT attrelid FROM pg_attribute
                        WHERE attrelid = (
                            SELECT oid FROM pg_class WHERE relname = table_rec.relname
                                AND relnamespace = (SELECT ns.oid FROM pg_namespace ns WHERE ns.nspname = p_schema_name)
                        ) AND attname='tableoid'
                    )
        LOOP
            v_table_ddl:=v_table_ddl||','||chr(10);
            v_table_ddl:=v_table_ddl||'CONSTRAINT '||constraint_rec.conname;
            v_table_ddl:=v_table_ddl||chr(10)||'    '||constraint_rec.constrainddef;
            firstrec := FALSE;
        END LOOP;
        v_table_ddl:=v_table_ddl||');';
        RETURN NEXT v_table_ddl;
    END LOOP;
END;
$BODY$
  LANGUAGE plpgsql VOLATILE
  COST 100;
  `.trim();
  await dbConnection.query({ sql });
};
