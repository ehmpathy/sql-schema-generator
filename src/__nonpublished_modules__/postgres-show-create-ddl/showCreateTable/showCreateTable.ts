import { DatabaseConnection } from './types';

export const showCreateTable = async ({
  dbConnection,
  schema,
  table,
}: {
  dbConnection: DatabaseConnection;
  schema: string;
  table: string;
}) => {
  // grab the ddl our function generates
  const result = await dbConnection.query({
    sql: 'SELECT public.show_create_table($1, $2) as ddl',
    values: [schema, table],
  });
  if (!result.rows)
    throw new Error(`could not find table '${schema}.${table}'`);
  return result.rows[0].ddl;
};
