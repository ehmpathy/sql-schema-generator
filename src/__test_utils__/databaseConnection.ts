import { Client, QueryResult } from 'pg';

export interface DatabaseConnection {
  query: (args: { sql: string; values?: (string | number)[] }) => Promise<QueryResult<any>>;
  end: () => Promise<void>;
}
export const getDatabaseConnection = async (): Promise<DatabaseConnection> => {
  const client = new Client({
    // hardcoded since this is just the integration test db config; settings configured in the provisioning/.../docker.file
    host: 'localhost',
    user: 'postgres',
    password: 'a-secure-password', // its just a docker integration test db, no worries about hardcoding this
    database: 'superimportantdb',
    port: 7821,
  });
  await client.connect();
  const connection = {
    query: ({ sql, values }: { sql: string; values?: (string | number)[] }) => client.query(sql, values),
    end: () => client.end(),
  };
  return connection;
};
