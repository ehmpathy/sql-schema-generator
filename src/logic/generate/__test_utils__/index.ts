export * from '../../../__test_utils__/databaseConnection';
export * from './createTablesForEntity';
export * from './dropTablesForEntity';
