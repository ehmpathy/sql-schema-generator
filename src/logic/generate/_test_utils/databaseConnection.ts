import mysql, { Connection, ConnectionOptions } from 'mysql2/promise';

export const getDatabaseConnection = async (): Promise<Connection> => {
  const dbConfig: ConnectionOptions = {
    // hardocded since this is just the integration test db config; settings configured in the provisioning/.../docker.file
    host: 'localhost',
    user: 'root',
    password: 'a-secure-password', // its just a docker integration test db, no worries about hardcoding this
    database: 'superimportantdb',
    port: 12821,
  };
  const connection = await mysql.createConnection(dbConfig);
  return connection;
};

export { Connection as DatabaseConnection };
