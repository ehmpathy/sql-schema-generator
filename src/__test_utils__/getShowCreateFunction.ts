import { normalizeCreateFunctionDdl } from '../__nonpublished_modules__/postgres-show-create-ddl/showCreateFunction/normalizeCreateFunctionDdl';
import { showCreateFunction } from '../__nonpublished_modules__/postgres-show-create-ddl/showCreateFunction/showCreateFunction';
import { DatabaseConnection } from './databaseConnection';

export const getShowCreateFunction = async ({
  func,
  dbConnection,
}: {
  func: string;
  dbConnection: DatabaseConnection;
}) => {
  const ddl = await showCreateFunction({ dbConnection, schema: 'public', func });
  return normalizeCreateFunctionDdl({ ddl })
    .replace(/public./g, '') // also, remove the schema namespace ('public' in this case)
    .trim();
};
