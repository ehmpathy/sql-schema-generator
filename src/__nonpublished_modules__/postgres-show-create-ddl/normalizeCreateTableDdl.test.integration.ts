import { DatabaseConnection, getDatabaseConnection } from '../../__test_utils__/databaseConnection';
import { normalizeCreateTableDdl } from './normalizeCreateTableDdl';
import { provisionShowCreateTableFunction } from './provisionShowCreateTableFunction';
import { showCreateTable } from './showCreateTable';

describe('showCreateTable', () => {
  let dbConnection: DatabaseConnection;
  beforeAll(async () => {
    dbConnection = await getDatabaseConnection();
  });
  afterAll(async () => {
    await dbConnection.end();
  });
  it('should be possible to get create statement of table', async () => {
    await provisionShowCreateTableFunction({ dbConnection });
    await dbConnection.query({ sql: 'DROP TABLE IF EXISTS test_tb_for_show_create_on;' });
    await dbConnection.query({
      sql: `
CREATE TABLE test_tb_for_show_create_on (
  id BIGSERIAL PRIMARY KEY,
  name VARCHAR(150),
  level VARCHAR(50) CHECK (level IN ('info', 'warn', 'error')) DEFAULT 'info',
  description TEXT NOT NULL
)
    `.trim(),
    });
    const ddl = await showCreateTable({ dbConnection, schema: 'public', table: 'test_tb_for_show_create_on' });
    const normalizedDdl = await normalizeCreateTableDdl({ ddl });
    expect(normalizedDdl).toMatchSnapshot();
  });
});
