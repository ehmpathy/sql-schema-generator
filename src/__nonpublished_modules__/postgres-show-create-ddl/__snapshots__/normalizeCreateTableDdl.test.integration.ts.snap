// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`showCreateTable should be possible to get create statement of table 1`] = `
"CREATE TABLE public.test_tb_for_show_create_on (
  id bigserial NOT NULL,
  name varchar(150) NULL,
  level varchar(50) NULL,
  CONSTRAINT test_tb_for_show_create_on_level_check CHECK (
    (
      (level) = ANY (
        (
          ARRAY ['info', 'warn', 'error']
        )
      )
    )
  ),
  CONSTRAINT test_tb_for_show_create_on_pkey PRIMARY KEY (id)
);"
`;
